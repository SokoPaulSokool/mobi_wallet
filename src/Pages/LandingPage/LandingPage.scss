.landing-page {
  .totalized {
    display: flex;
    align-items: center;
    margin: 0 auto;
    width: 100%;
  }

  .profile {
    position: fixed;
    display: flex;
    flex-direction: column;
    top: 10px;
    right: 10px;
    background: white;
    z-index: 2;
    &-details{
      display: flex;
      align-items: center;
    }
    .circle {
      width: 60px;
      height: 60px;
      margin-left: 10px;
      border-radius: 50%;
      display: flex;
      align-items: center;
      justify-content: center;
    }
  }

  .currencies{
    position: relative;
    margin-top: 6vh;
      height: 60vh;
      width: 100%;
      display: flex;
      flex-wrap: wrap;
      overflow-y: auto;
  }
}
