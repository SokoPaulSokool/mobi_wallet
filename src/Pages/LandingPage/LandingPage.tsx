import React, { useEffect } from "react";
import { useContext, useState } from "react";
import CurrencyCard from "../../components/CurrencyCard/CurrencyCard";
import { GlobalContext } from "../../context/provider";
import "./LandingPage.scss";
import ExchangeCurrencyDialog from "../../components/Dialogs/ExchangeCurrencyDialog/ExchangeCurrencyDialog";
import DepositCurrencyDialog from "../../components/Dialogs/DepositCurrencyDialog/DepositCurrencyDialog";
import FloatingMenu from "../../components/FloatingMenu/FloatingMenu";
import {
  TRANSACTION_HISTORY,
  ADD_NEW_CURRENCY,
  CHANGE_DEFAULT_CURRENCY,
} from "../../constants/generalConstants";
import AddCurrencyDialog from "../../components/Dialogs/AddCurrencyDialog/AddCurrencyDialog";
import TransactionHistoryDialog from "../../components/Dialogs/TransactionHistoryDialog/TransactionHistoryDialog";
import Currency from "../../interfaces/CurrencyInterfaces";
import {
  calculateTotalisedAmount,
  roundNumber,
} from "../../helpers/generalHelpers";
import { Autocomplete } from "@material-ui/lab";
import { TextField } from "@material-ui/core";
import { changeDefaultCurrency } from "../../context/actions/currenciesActions";

const LandingPage: React.FC = () => {
  const { currenciesState, currenciesDispatch } = useContext(GlobalContext);
  const [currencyList, setCurrencyList] = useState<Currency[]>([]);
  const [selectedCurrency, setSelectedCurrency] = useState<
    Currency | undefined
  >(undefined);

  const [totalisedAmount, setTotalisedAmount] = useState<number | null>(null);
  const [defaultCurrency, setDefaultCurrency] = useState<Currency | null>(null);

  const [isCurrencyExchangeDialogOpen, setIsCurrencyExchangeDialogOpen] =
    useState(false);
  const [isDepositCurrencyDialogOpen, setIsDepositCurrencyDialogOpen] =
    useState(false);
  const [isAddNewCurrencyDialogOpen, setIsAddNewCurrencyDialogOpen] =
    useState(false);
  const [isTransactionHistoryDialogOpen, setIsTransactionHistoryDialogOpen] =
    useState(false);


    useEffect(() => {
      if (currenciesState?.currencies) {
        const currencies = Object.values(currenciesState?.currencies).map(
          (currency) => currency
        );
        setCurrencyList(currencies);
      }
      return () => {
      };
    }, []);

  useEffect(() => {
    if (currenciesState?.currencies) {
      const currencies = Object.values(currenciesState?.currencies).map(
        (currency) => currency
      );
      setCurrencyList(currencies);
    }
    return () => {};
  }, [currenciesState?.currencies]);  

  const updateTotalisedAmount = () => {
    if (currenciesState?.defaultCurrency && currencyList) {
      const total = roundNumber(
        calculateTotalisedAmount(currencyList, currenciesState?.defaultCurrency)
      );
      setTotalisedAmount(total);
    }
  };

  useEffect(() => {
    if (currenciesState?.defaultCurrency) {
      updateTotalisedAmount();
      setDefaultCurrency(currenciesState?.defaultCurrency);
    }
    return () => {};
  }, [currenciesState?.defaultCurrency, currencyList]);

  const handleFloatingMenuClick = (clickedOption: String) => {
    switch (clickedOption) {
      case TRANSACTION_HISTORY:
        setIsTransactionHistoryDialogOpen(!isTransactionHistoryDialogOpen);

        break;
      case ADD_NEW_CURRENCY:
        setIsAddNewCurrencyDialogOpen(!isAddNewCurrencyDialogOpen);
        break;
      default:
        break;
    }
  };

  const handleCurrencyChange = (currency: Currency | null) => {
    if (currency) {
      setDefaultCurrency(currency);
      changeDefaultCurrency(currency)(currenciesDispatch);
    }
  };

  return (
    <div data-testid="landing-page" className="landing-page container mt-2">
      <ExchangeCurrencyDialog
        open={isCurrencyExchangeDialogOpen}
        selectedValue={selectedCurrency}
        currencies={currencyList}
        onClose={() => {
          setIsCurrencyExchangeDialogOpen(false);
        }}
      ></ExchangeCurrencyDialog>
      <DepositCurrencyDialog
        open={isDepositCurrencyDialogOpen}
        selectedValue={selectedCurrency}
        onClose={() => {
          setIsDepositCurrencyDialogOpen(false);
        }}
      ></DepositCurrencyDialog>
      <AddCurrencyDialog
        open={isAddNewCurrencyDialogOpen}
        currencies={currencyList}
        onClose={() => {
          setIsAddNewCurrencyDialogOpen(false);
        }}
      ></AddCurrencyDialog>
      <TransactionHistoryDialog
        open={isTransactionHistoryDialogOpen}
        onClose={() => {
          setIsTransactionHistoryDialogOpen(false);
        }}
      ></TransactionHistoryDialog>

      <h1 data-testid="app-name">Mobi Wallet</h1>
      <div className="profile">
        <div className="profile-details">
          <h3 data-testid="user-name">John Doe</h3>
          <div className="circle app-bg-accent"></div>
        </div>
        <div className="default-currency">
          <Autocomplete
            data-testid="default-currency"
            className="my-2 w-100 px-0"
            options={currencyList ? currencyList : []}
            value={defaultCurrency}
            getOptionLabel={(option) => option.units}
            onChange={(event, newValue) => handleCurrencyChange(newValue)}
            renderInput={(params) => (
              <TextField {...params} label="Default Currency" />
            )}
          />
        </div>
      </div>
      <div className="totalized">
        <p data-testid="totallised-value-label">Totalised value</p>
        <h1 data-testid="totallised-value-details">
          {currenciesState?.defaultCurrency.units} {totalisedAmount}
        </h1>
      </div>
      <div className="currencies" data-testid="currency-cards">
        {currencyList.map((currency: Currency, i: number) => {
          return (
            <CurrencyCard
              key={i + "p"}
              currency={currency}
              onDepositClick={() => {
                setIsDepositCurrencyDialogOpen(true);
                setSelectedCurrency(currency);
              }}
              onExchangeClick={() => {
                setIsCurrencyExchangeDialogOpen(true);
                setSelectedCurrency(currency);
              }}
            ></CurrencyCard>
          );
        })}
      </div>
      <FloatingMenu
        currentlySelectedMenu={""}
        onMenuClick={handleFloatingMenuClick}
      ></FloatingMenu>
    </div>
  );
};

export default LandingPage;
