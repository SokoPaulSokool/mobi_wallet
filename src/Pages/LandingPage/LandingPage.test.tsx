import React from "react";
import { render, fireEvent } from "@testing-library/react";
import LandingPage from "./LandingPage";
import GlobalProvider from "../../context/provider";

test("renders LandingPage", () => {
  const { getByTestId } = render(<LandingPage />);
  const appNameElement = getByTestId("app-name");
  const totalisedValueLabel = getByTestId("totallised-value-label");
  const totalisedValueDetails = getByTestId("totallised-value-details");

  expect(appNameElement).toBeInTheDocument();
  expect(totalisedValueLabel).toBeInTheDocument();
  expect(totalisedValueDetails).toBeInTheDocument();
});
test("should have expected text", () => {
  const { getByTestId } = render(<LandingPage />);
  const appNameElement = getByTestId("app-name");
  const totalisedValueLabel = getByTestId("totallised-value-label");
  const userName = getByTestId("user-name");

  expect(appNameElement.textContent).toBe("Mobi Wallet");
  expect(totalisedValueLabel.textContent).toBe("Totalised value");
  expect(userName.textContent).toBe("John Doe");
});

test("should change default currency", async () => {
  const { getByTestId } = render(
    <GlobalProvider>
      <LandingPage />
    </GlobalProvider>
  );

  const autocomplete = getByTestId("default-currency");
  const defaultCurrency = autocomplete.querySelector("input");
  autocomplete.focus();
  if (defaultCurrency) {
    await fireEvent.change(defaultCurrency, { target: { value: "us" } });
    await fireEvent.keyDown(autocomplete, { key: "ArrowDown" });
    await fireEvent.keyDown(autocomplete, { key: "Enter" });
  }

  expect(defaultCurrency.value).toBe("USD");
});

test("currency-card button handle deposit and exchange", async () => {
  const { getByTestId } = render(
    <GlobalProvider>
      <LandingPage />
    </GlobalProvider>
  );

  const card = getByTestId("currency-cards");
  const cardButtons = card.querySelectorAll("button");

  if (cardButtons) {
    fireEvent.click(cardButtons[0]);
    fireEvent.click(cardButtons[1]);
  }
});

test("floating menu", async () => {
  const { getByTestId, asFragment } = render(
    <GlobalProvider>
      <LandingPage />
    </GlobalProvider>
  );

  const landingPage = getByTestId("landing-page");
  const transactionHistory = landingPage
    .getElementsByClassName("floating-menu")
    .item(0)
    ?.getElementsByClassName("item-details")
    .item(0);
  const addNewCurrency = landingPage
    .getElementsByClassName("floating-menu")
    .item(0)
    ?.getElementsByClassName("item-details")
    .item(1);

  if (transactionHistory) {
    fireEvent.click(transactionHistory);
  }
  if (addNewCurrency) {
    fireEvent.click(addNewCurrency);
  }
});
