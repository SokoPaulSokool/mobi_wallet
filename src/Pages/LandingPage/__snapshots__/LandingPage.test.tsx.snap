// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`floating menu 1`] = `
<DocumentFragment>
  <div
    class="landing-page container mt-2"
    data-testid="landing-page"
  >
    <h1
      data-testid="app-name"
    >
      Mobi Wallet
    </h1>
    <div
      class="profile"
    >
      <div
        class="profile-details"
      >
        <h3
          data-testid="user-name"
        >
          John Doe
        </h3>
        <div
          class="circle app-bg-accent"
        />
      </div>
      <div
        class="default-currency"
      >
        <div
          aria-expanded="false"
          class="MuiAutocomplete-root my-2 w-100 px-0 MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon"
          data-testid="default-currency"
          role="combobox"
        >
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
              data-shrink="true"
              for="mui-9912"
              id="mui-9912-label"
            >
              Default Currency
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
            >
              <input
                aria-autocomplete="list"
                aria-invalid="false"
                autocapitalize="none"
                autocomplete="off"
                class="MuiInputBase-input MuiInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd"
                id="mui-9912"
                spellcheck="false"
                type="text"
                value="USD"
              />
              <div
                class="MuiAutocomplete-endAdornment"
              >
                <button
                  aria-label="Clear"
                  class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator MuiAutocomplete-clearIndicatorDirty"
                  tabindex="-1"
                  title="Clear"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
                <button
                  aria-label="Open"
                  class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator"
                  tabindex="-1"
                  title="Open"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M7 10l5 5 5-5z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="totalized"
    >
      <p
        data-testid="totallised-value-label"
      >
        Totalised value
      </p>
      <h1
        data-testid="totallised-value-details"
      >
        USD 754.62
      </h1>
    </div>
    <div
      class="currencies"
      data-testid="currency-cards"
    >
      <div
        class="currency-card app-bg-primary app-clr-white"
      >
        <div
          class="close-btn"
        >
          Close
        </div>
        <div
          class="metal-plate app-bg-accent"
        >
          <div
            class="line"
          />
          <div
            class="line"
          />
          <div
            class="line"
          />
        </div>
        <div
          class="details"
        >
          <div
            class="labels"
            data-testid="labels"
          >
            <p
              class=""
            >
              Currency:
            </p>
            <h3
              class=""
            >
              Balance:
            </h3>
          </div>
          <div
            class="values"
          >
            <p
              class=""
              data-testid="card-units"
            >
              EUR
            </p>
            <h3
              class=""
              data-testid="card-amount"
            >
              500
            </h3>
          </div>
        </div>
        <div
          class="buttons"
          data-testid="card-buttons"
        >
          <button
            class="app-button app-bg-white app-clr-accent"
            data-testid="app-button"
            type="button"
          >
            Deposit
          </button>
          <button
            class="app-button app-bg-white app-clr-black"
            data-testid="app-button"
            type="button"
          >
            Exchange
          </button>
        </div>
      </div>
      <div
        class="currency-card app-bg-primary app-clr-white"
      >
        <div
          class="close-btn"
        >
          Close
        </div>
        <div
          class="metal-plate app-bg-accent"
        >
          <div
            class="line"
          />
          <div
            class="line"
          />
          <div
            class="line"
          />
        </div>
        <div
          class="details"
        >
          <div
            class="labels"
            data-testid="labels"
          >
            <p
              class=""
            >
              Currency:
            </p>
            <h3
              class=""
            >
              Balance:
            </h3>
          </div>
          <div
            class="values"
          >
            <p
              class=""
              data-testid="card-units"
            >
              USD
            </p>
            <h3
              class=""
              data-testid="card-amount"
            >
              100
            </h3>
          </div>
        </div>
        <div
          class="buttons"
          data-testid="card-buttons"
        >
          <button
            class="app-button app-bg-white app-clr-accent"
            data-testid="app-button"
            type="button"
          >
            Deposit
          </button>
          <button
            class="app-button app-bg-white app-clr-black"
            data-testid="app-button"
            type="button"
          >
            Exchange
          </button>
        </div>
      </div>
      <div
        class="currency-card app-bg-primary app-clr-white"
      >
        <div
          class="close-btn"
        >
          Close
        </div>
        <div
          class="metal-plate app-bg-accent"
        >
          <div
            class="line"
          />
          <div
            class="line"
          />
          <div
            class="line"
          />
        </div>
        <div
          class="details"
        >
          <div
            class="labels"
            data-testid="labels"
          >
            <p
              class=""
            >
              Currency:
            </p>
            <h3
              class=""
            >
              Balance:
            </h3>
          </div>
          <div
            class="values"
          >
            <p
              class=""
              data-testid="card-units"
            >
              CHF
            </p>
            <h3
              class=""
              data-testid="card-amount"
            >
              30
            </h3>
          </div>
        </div>
        <div
          class="buttons"
          data-testid="card-buttons"
        >
          <button
            class="app-button app-bg-white app-clr-accent"
            data-testid="app-button"
            type="button"
          >
            Deposit
          </button>
          <button
            class="app-button app-bg-white app-clr-black"
            data-testid="app-button"
            type="button"
          >
            Exchange
          </button>
        </div>
      </div>
      <div
        class="currency-card app-bg-primary app-clr-white"
      >
        <div
          class="close-btn"
        >
          Close
        </div>
        <div
          class="metal-plate app-bg-accent"
        >
          <div
            class="line"
          />
          <div
            class="line"
          />
          <div
            class="line"
          />
        </div>
        <div
          class="details"
        >
          <div
            class="labels"
            data-testid="labels"
          >
            <p
              class=""
            >
              Currency:
            </p>
            <h3
              class=""
            >
              Balance:
            </h3>
          </div>
          <div
            class="values"
          >
            <p
              class=""
              data-testid="card-units"
            >
              UGX
            </p>
            <h3
              class=""
              data-testid="card-amount"
            >
              20
            </h3>
          </div>
        </div>
        <div
          class="buttons"
          data-testid="card-buttons"
        >
          <button
            class="app-button app-bg-white app-clr-accent"
            data-testid="app-button"
            type="button"
          >
            Deposit
          </button>
          <button
            class="app-button app-bg-white app-clr-black"
            data-testid="app-button"
            type="button"
          >
            Exchange
          </button>
        </div>
      </div>
      <div
        class="currency-card app-bg-primary app-clr-white"
      >
        <div
          class="close-btn"
        >
          Close
        </div>
        <div
          class="metal-plate app-bg-accent"
        >
          <div
            class="line"
          />
          <div
            class="line"
          />
          <div
            class="line"
          />
        </div>
        <div
          class="details"
        >
          <div
            class="labels"
            data-testid="labels"
          >
            <p
              class=""
            >
              Currency:
            </p>
            <h3
              class=""
            >
              Balance:
            </h3>
          </div>
          <div
            class="values"
          >
            <p
              class=""
              data-testid="card-units"
            >
              KES
            </p>
            <h3
              class=""
              data-testid="card-amount"
            >
              50
            </h3>
          </div>
        </div>
        <div
          class="buttons"
          data-testid="card-buttons"
        >
          <button
            class="app-button app-bg-white app-clr-accent"
            data-testid="app-button"
            type="button"
          >
            Deposit
          </button>
          <button
            class="app-button app-bg-white app-clr-black"
            data-testid="app-button"
            type="button"
          >
            Exchange
          </button>
        </div>
      </div>
    </div>
    <div
      class="floating-menu"
      data-testid="floating-menu-list"
    >
      <div
        class="menu-items"
      >
        <div
          class="item"
          data-testid="floating-menu-list-item-0"
        >
          <p>
            Transaction History
          </p>
          <div
            class="circle"
          />
        </div>
        <div
          class="item"
          data-testid="floating-menu-list-item-1"
        >
          <p>
            Add New Currency
          </p>
          <div
            class="circle"
          />
        </div>
      </div>
      <div
        class="circle menu-toggle"
        data-testid="floating-menu-list-toggle"
      />
    </div>
  </div>
</DocumentFragment>
`;
